File: app/src/androidTest/java/org/isoron/uhabits/activities/habits/list/views/CheckmarkPanelViewTest.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.activities.habits.list.views;

import android.support.test.runner.AndroidJUnit4;
import android.test.suitebuilder.annotation.*;

import org.isoron.uhabits.models.Checkmark;
import org.isoron.uhabits.models.Habit;
import org.isoron.uhabits.BaseViewTest;
import org.isoron.uhabits.utils.ColorUtils;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.concurrent.CountDownLatch;

@RunWith(AndroidJUnit4.class)
@MediumTest
public class CheckmarkPanelViewTest extends BaseViewTest
{
    public static final String PATH = "habits/list/CheckmarkPanelView/";

    private CountDownLatch latch;
    private CheckmarkPanelView view;
    private int checkmarks[];

    @Override
    @Before
    public void setUp()
    {
        super.setUp();
        setSimilarityCutoff(0.03f);
        prefs.setShouldReverseCheckmarks(false);

        Habit habit = fixtures.createEmptyHabit();

        latch = new CountDownLatch(1);
        checkmarks = new int[]{
            Checkmark.CHECKED_EXPLICITLY, Checkmark.UNCHECKED,
                Checkmark.CHECKED_IMPLICITLY, Checkmark.CHECKED_EXPLICITLY};

        view = new CheckmarkPanelView(targetContext);
        view.setHabit(habit);
        view.setCheckmarkValues(checkmarks);
        view.setButtonCount(4);
        view.setColor(ColorUtils.getAndroidTestColor(7));

        measureView(view, dpToPixels(200), dpToPixels(200));
    }

//    protected void waitForLatch() throws InterruptedException
//    {
//        assertTrue("Latch timeout", latch.await(1, TimeUnit.SECONDS));
//    }

    @Test
    public void testRender() throws Exception
    {
        assertRenders(view, PATH + "render.png");
    }

//    @Test
//    public void testToggleCheckmark_withLeftToRight() throws Exception
//    {
//        setToggleListener();
//        view.getButton(1).performToggle();
//        waitForLatch();
//    }
//
//    @Test
//    public void testToggleCheckmark_withReverseCheckmarks() throws Exception
//    {
//        prefs.setShouldReverseCheckmarks(true);
//        view.setCheckmarkValues(checkmarks); // refresh after preference change
//
//        setToggleListener();
//        view.getButton(2).performToggle();
//        waitForLatch();
//    }
}

File: uhabits-android/src/androidTest/java/org/isoron/uhabits/activities/habits/list/views/CheckmarkPanelViewTest.kt
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.activities.habits.list.views

import android.support.test.filters.*
import android.support.test.runner.*
import org.hamcrest.CoreMatchers.*
import org.hamcrest.MatcherAssert.*
import org.isoron.uhabits.*
import org.isoron.uhabits.core.models.Checkmark.*
import org.isoron.uhabits.utils.*
import org.junit.*
import org.junit.runner.*

@RunWith(AndroidJUnit4::class)
@MediumTest
class CheckmarkPanelViewTest : BaseViewTest() {

    private val PATH = "habits/list/CheckmarkPanelView"
    private lateinit var view: CheckmarkPanelView

    @Before
    override fun setUp() {
        super.setUp()
        prefs.isCheckmarkSequenceReversed = false

        val checkmarks = intArrayOf(CHECKED_EXPLICITLY,
                                    CHECKED_EXPLICITLY,
                                    CHECKED_IMPLICITLY,
                                    UNCHECKED,
                                    UNCHECKED,
                                    UNCHECKED,
                                    CHECKED_EXPLICITLY)

        view = component.getCheckmarkPanelViewFactory().create().apply {
            values = checkmarks
            buttonCount = 4
            color = PaletteUtils.getAndroidTestColor(7)
        }
        view.onAttachedToWindow()
        measureView(view, dpToPixels(200), dpToPixels(200))
    }

    @After
    public override fun tearDown() {
//        view.onDetachedFromWindow()
        super.tearDown()
    }

    @Test
    fun testRender() {
        assertRenders(view, "$PATH/render.png")
    }

    @Test
    fun testRender_withDifferentColor() {
        view.color = PaletteUtils.getAndroidTestColor(1)
        assertRenders(view, "$PATH/render_different_color.png")
    }

    @Test
    fun testRender_Reversed() {
        prefs.isCheckmarkSequenceReversed = true
        assertRenders(view, "$PATH/render_reversed.png")
    }

    @Test
    fun testRender_withOffset() {
        view.dataOffset = 3
        assertRenders(view, "$PATH/render_offset.png")
    }

    @Test
    fun testToggle() {
        var timestamps = mutableListOf<Long>()
        view.onToggle = { timestamps.add(it) }
        view.buttons[0].performLongClick()
        view.buttons[2].performLongClick()
        view.buttons[3].performLongClick()
        assertThat(timestamps, equalTo(listOf(day(0), day(2), day(3))))
    }

    @Test
    fun testToggle_withOffset() {
        var timestamps = LongArray(0)
        view.dataOffset = 3
        view.onToggle = { timestamps += it }
        view.buttons[0].performLongClick()
        view.buttons[2].performLongClick()
        view.buttons[3].performLongClick()
        assertThat(timestamps, equalTo(longArrayOf(day(3), day(5), day(6))))
    }
}